### PR TITLE
fix: normalize behavior of `win.setOpacity()` for invalid number values across operating systems

### PR DESCRIPTION
**Pull request by erickzhao**
### Description of Change
Small fix to normalize behavior for invalid values for the [`win.setOpacity()`](https://electronjs.org/docs/api/browser-window#winsetopacityopacity-windows-macos).

For example: `win.setOpacity(100)` or `win.setOpacity(-100)`.

#### Current behavior with invalid values
* Windows:
  * `win.setOpacity(opacity)`: supposed to normalize the [0,1] opacity values to 255. However, for invalid values, passing in `opacity * 255` leads to unexpected behavior.
  * `win.getOpacity()`: retrieves the invalid `opacity` value that was previously passed in.
* macOS:
  * `win.setOpacity(opacity)`: bounds opacity value to the [0,1] range (i.e. `(opacity < 0) -> (opacity = 0)` and `(opacity > 1) -> (opacity = 1)`)
  * `win.getOpacity()`: retrieves the invalid `opacity` value that was previously passed in.

* Linux (unsupported):
  * `win.setOpacity(opacity)`: takes no action on the actual window opacity, but stores the `opacity` value.
  * `win.getOpacity()`: retrieves the previously stored `opacity` value.

#### Proposed behavior

* macOS/Windows:
  * `win.setOpacity(opacity)`: bounds opacity value to the [0,1] range
  * `win.getOpacity()`: retrieves the bounded opacity value.
* Linux:
  * `win.setOpacity(opacity)`:  takes no action.
  * `win.getOpacity()`: always returns 1.

### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

### Release Notes

Notes: Normalized out-of-bound value behavior for the `setOpacity()` API in `BrowserWindow`.

### CODE REVIEW

File: spec-main/api-browser-window-spec.ts [hunk 1]
```diff
@@ -1241,6 +1241,12 @@ describe('BrowserWindow module', () => {
   })
 
   describe('BrowserWindow.setOpacity(opacity)', () => {
+    before(function () {
```
Comment (codebytere) [inline spec-main/api-browser-window-spec.ts]: Use  the `ifit` helper here

File: shell/browser/native_window_views.cc [hunk 2]
```diff
@@ -880,16 +880,25 @@ bool NativeWindowViews::HasShadow() {
...
     LONG ex_style = ::GetWindowLong(hwnd, GWL_EXSTYLE);
     ex_style |= WS_EX_LAYERED;
     ::SetWindowLong(hwnd, GWL_EXSTYLE, ex_style);
     layered_ = true;
   }
-  ::SetLayeredWindowAttributes(hwnd, 0, opacity * 255, LWA_ALPHA);
+  ::SetLayeredWindowAttributes(hwnd, 0, boundedOpacity * 255, LWA_ALPHA);
+  opacity_ = boundedOpacity;
+#else
+  opacity_ = 1;  // setOpacity unsupported on Linux
```
Comment (ckerr) [inline shell/browser/native_window_views.cc]:
```c++
    layered_ = true;
  }

  opacity_ = base::ClampToRange(opacity, 0, 1);
  ::SetLayeredWindowAttributes(hwnd, 0, boundedOpacity * 255, LWA_ALPHA);
#else
  opacity_ = 1.0;
#endif

File: spec-main/api-browser-window-spec.ts [hunk 3]
```diff
@@ -1257,6 +1256,14 @@ describe('BrowserWindow module', () => {
         expect(w.getOpacity()).to.equal(1.0)
       }).to.not.throw()
     })
+
+    it('does not change opacity if given number out of bounds', () => {
```
Comment (ckerr) [inline spec-main/api-browser-window-spec.ts]: Well, we _do_ change opacity if the number is out of bounds, we just bring it in bounds first

```suggestion
    it('clamps opacity to [0.0...1.0]', () => {
```

File: shell/browser/native_window_mac.mm [hunk 4]
```diff
@@ -1032,7 +1032,14 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
 }
 
 void NativeWindowMac::SetOpacity(const double opacity) {
-  [window_ setAlphaValue:opacity];
+  double boundedOpacity = opacity;
+  if (opacity < 0) {
+    boundedOpacity = 0;
+  } else if (opacity > 1) {
+    boundedOpacity = 1;
+  }
```
[the remark this answers is not in the document]
Comment (erickzhao) [reply shell/browser/native_window_mac.mm]: Very useful utility function that I didn't know about! Thanks for the heads-up. 😄

File: spec-main/api-browser-window-spec.ts [hunk 5]
```diff
@@ -1240,7 +1238,8 @@ describe('BrowserWindow module', () => {
     })
   })
 
-  describe('BrowserWindow.setOpacity(opacity)', () => {
+
+  ifdescribe(process.platform !== 'linux')('BrowserWindow.setOpacity(opacity)', () => {
```
Comment (codebytere) [inline spec-main/api-browser-window-spec.ts]: Given that this change also normalizes the result of `getOpacity` on Linux to 1, we should also test for that!

File: shell/browser/native_window_mac.mm [hunk 6]
```diff
@@ -1032,7 +1033,8 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
 }
 
 void NativeWindowMac::SetOpacity(const double opacity) {
-  [window_ setAlphaValue:opacity];
+  double boundedOpacity = base::ClampToRange(opacity, 0.0, 1.0);
```
Comment (ckerr) [inline shell/browser/native_window_mac.mm]: _(nit)_

```suggestion
  const double boundedOpacity = base::ClampToRange(opacity, 0.0, 1.0);
```

File: shell/browser/native_window_views.cc [hunk 7]
```diff
@@ -880,16 +881,19 @@ bool NativeWindowViews::HasShadow() {
 
 void NativeWindowViews::SetOpacity(const double opacity) {
 #if defined(OS_WIN)
+  double boundedOpacity = base::ClampToRange(opacity, 0.0, 1.0);
```
Comment (ckerr) [inline shell/browser/native_window_views.cc]: _(nit)_

```suggestion
  const double boundedOpacity = base::ClampToRange(opacity, 0.0, 1.0);
```

File: spec-main/api-browser-window-spec.ts [hunk 8]
```diff
@@ -1240,22 +1238,44 @@ describe('BrowserWindow module', () => {
...
+        w.setOpacity(-100)
         expect(w.getOpacity()).to.equal(0.0)
+      })
+    })
+
+    ifdescribe(process.platform === 'linux')(('Linux'), () => {
+      it('sets 1 regardless of parameter', () => {
+        const w = new BrowserWindow({ show: false })
+        w.setOpacity(0)
+        expect(w.getOpacity()).to.equal(1.0)
```
Comment (ckerr) [inline spec-main/api-browser-window-spec.ts]: :+1: on the tests

**Comment by codebytere** [comment]
@erickzhao once conflicts are handled this should be g2g!

**Comment by erickzhao** [comment]
@codebytere done! 🚀

**Comment by codebytere** [comment]
@erickzhao can you please handle these manual backports when you have a moment?

**Pull request merged**